samza-operator/src/main/java/org/apache/samza/operators/spec/OperatorSpec.java

```diff
@@ -18,20 +18,45 @@
  */
 package org.apache.samza.operators.spec;
 
-import org.apache.samza.operators.MessageStream;
-import org.apache.samza.operators.data.MessageEnvelope;
+import org.apache.samza.annotation.InterfaceStability;
+import org.apache.samza.config.Config;
+import org.apache.samza.operators.MessageStreamImpl;
+import org.apache.samza.task.TaskContext;
 
 
 /**
  * A stateless serializable stream operator specification that holds all the information required
- * to transform the input {@link MessageStream} and produce the output {@link MessageStream}.
+ * to transform the input {@link MessageStreamImpl} and produce the output {@link MessageStreamImpl}.
+ *
+ * @param <OM>  the type of output message from the operator
  */
-public interface OperatorSpec<OM extends MessageEnvelope> {
+@InterfaceStability.Unstable
+public interface OperatorSpec<OM> {
+
+  enum OpCode {
```

Comment:
For Xinyu's patch, he needs to get the specific operator code when inspecting the logic DAG to find "partitionBy" that is going to trigger creation of intermediate topics. It would also make sense if we want to visualize / serialize the DAG, or apply some optimization rules in the DAG.